quickwit-janitor/src/actors/janitor_service.rs

```diff
@@ -17,48 +17,84 @@
 // You should have received a copy of the GNU Affero General Public License
 // along with this program. If not, see <http://www.gnu.org/licenses/>.
 
-use std::path::PathBuf;
 use std::sync::Arc;
 use std::time::Duration;
 
 use async_trait::async_trait;
-use quickwit_actors::{Actor, ActorContext, ActorExitStatus, Handler};
-use quickwit_metastore::Metastore;
-use quickwit_storage::StorageUriResolver;
-use tracing::info;
-
-const RUN_INTERVAL: Duration = if cfg!(test) {
-    Duration::from_secs(60) // 1min
-} else {
-    Duration::from_secs(60 * 60) // 1h
+use quickwit_actors::{
+    Actor, ActorContext, ActorExitStatus, ActorHandle, Handler, Health, Supervisable,
 };
+use quickwit_metastore::{Metastore, MetastoreError};
+use quickwit_storage::{StorageResolverError, StorageUriResolver};
+use thiserror::Error;
+use tracing::{error, info};
+
+use super::garbage_collector::{GarbageCollector, GarbageCollectorCounters};
+
+const OBSERVATION_INTERVAL: Duration = Duration::from_secs(1);
+
+#[derive(Error, Debug)]
+pub enum JanitorServiceError {
+    #[error("Garbage collector already exists.")]
+    GarbageCollectionAlreadyExists,
+    #[error("Failed to resolve the storage `{0}`.")]
+    StorageError(#[from] StorageResolverError),
+    #[error("Metastore error `{0}`.")]
+    MetastoreError(#[from] MetastoreError),
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct SpawnGarbageCollector;
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Observe;
+
+#[derive(Clone, Copy, Debug, Default)]
+struct Supervise;
 
 #[derive(Clone, Debug, Default)]
 pub struct JanitorServiceCounters {
-    /// Number of passes.
-    pub num_passes: usize,
+    pub garbage_collector_counters: GarbageCollectorCounters,
 }
 
 pub struct JanitorService {
-    _data_dir_path: PathBuf,
-    _metastore: Arc<dyn Metastore>,
-    _storage_resolver: StorageUriResolver,
+    node_id: String,
+    metastore: Arc<dyn Metastore>,
+    storage_resolver: StorageUriResolver,
+    garbage_collector_handle_opt: Option<ActorHandle<GarbageCollector>>,
     counters: JanitorServiceCounters,
 }
 
 impl JanitorService {
     pub fn new(
-        data_dir_path: PathBuf,
+        node_id: String,
         metastore: Arc<dyn Metastore>,
         storage_resolver: StorageUriResolver,
     ) -> Self {
         Self {
-            _data_dir_path: data_dir_path,
-            _metastore: metastore,
-            _storage_resolver: storage_resolver,
+            node_id,
+            metastore,
+            storage_resolver,
+            garbage_collector_handle_opt: None,
             counters: JanitorServiceCounters::default(),
         }
     }
+
+    async fn spawn_garbage_collector(
```

Comment:
I would do that right in the new. I don't feel the need for a `SpawnGarbageCollector` message, we can do everything in the new method and we're done.